### ISSUE TITLE
add python 3 support

### ISSUE THREAD

**Issue by laliberte** [ESGF/esgf-pyclient#10]
Is there any plans to port this code to python 3? If you want, I could do it but it appears that the authentication might not be easily ported since it relies on MyProxyClient, which is also fairly out of date.

**Comment by agstephens** [comment]
Hi @laliberte, 

There are no plans to port to python 3. At present, most of our community is still rooted in python 2.7. It would be great if you have the time to update it.

Thanks, Ag

**Comment by laliberte** [comment]
Hi @agstephens, I tought the same until I realized recently that things seem to be moving pretty fast on the python 3 front with some key packages dropping features in python 2.7. So, if you don't mind, I would like to give a  jab at the python 3 migration.

Here's what I think would be the best way to go at it:

1. Reorganize the tests a bit and specify expected failures and skip tests that cannot succeed.
2. Ideally, travis-ci integration.
3. switch from using ``urllib2`` to using the ``requests`` library.

I have made some progress on 1 & 2. What do you think about 3? I'll create a PR later today as a basis of discussion for 1 & 2.

**Comment by agstephens** [comment]
Hi @laliberte, I have merged into master having re-run the tests. I have just set up Travis integration for the master branch - thanks for the nudges regarding good practice. I also appreciate your nice decorators to manage the logon details in the test suite. Much slicker. I agree that switching to using the `requests` would be a nice upgrade. Thanks, Ag

**Comment by philipkershaw** [comment]
Hi @laliberte,

I've made a dual Python 2/3 compatible version of MyProxyClient.  It's in devel at the moment:

https://github.com/cedadev/MyProxyClient/tree/devel 

should be able to put on PyPI v soon

**Comment by agstephens** [comment]
@philipkershaw, how is your py2/3 compatible version of MyProxyClient doing? Thanks, Ag

**Comment by philipkershaw** [comment]
It's completed but needs minting in PyPI

**Comment by philipkershaw** [comment]
New version 2.0 now in PyPI

**Comment by lewismc** [comment]
Hi folks, where is this effort currently at? We are looking to upgrade over at Apache Open Climate Workbench and eager to chip in if it would be of help. Thanks.

**Comment by philipkershaw** [comment]
Sorry for the slow response @lewismc, yes it would be good to have contributions.  The recent work has mostly been maintenance AFAIK

**Comment by lewismc** [comment]
OK doke, i think i will make an attempt to work on this issue then. @agoodm FYI.

**Comment by philipkershaw** [comment]
Thanks, the Python 3 port is pretty much finished in devel I think but it needs thorough testing.

**Comment by jhamman** [comment]
Does anyone have an update on this issue? It sounds like it is completed but waiting on some testing.

**Comment by lewismc** [comment]
IMHO master branch should be released and feedback should be posted here in issues. The source sitting in Git gathering dust is not really working as a mechanism for people testing the package. @philipkershaw @agstephens any comments?

**Comment by agstephens** [comment]
Hi all, I appreciate your feedback. Was looking at this last week and getting some outstanding tests working on Py2.7 and Py3.3. I'll cut a release and people can let us know if there are any problems.

**Comment by agstephens** [comment]
Closing this as we have released a Python 3 compatible-version at v0.2.1:

https://github.com/ESGF/esgf-pyclient/releases/tag/0.2.1

This release includes the following features:

 - Python 3 support for the library
 - `verify` option in LogonManager
 - Works with Python 3 version of MyProxyClient
 - Testing structure with `pytest` has been improved

This release is available on GitHub at:

https://github.com/ESGF/esgf-pyclient/releases/tag/0.2.1

And PyPI at:

https://pypi.python.org/pypi/esgf-pyclient 

You can typically install it with:

    `pip install esgf-pyclient`

The latest release is documented at:

http://esgf-pyclient.readthedocs.io/en/latest/
